numpy/core/src/multiarray/methods.c

```diff
@@ -1991,7 +1991,8 @@ array_setstate(PyArrayObject *self, PyObject *args)
     }
 
     if ((PyArray_FLAGS(self) & NPY_ARRAY_OWNDATA)) {
-        PyDataMem_FREE(PyArray_DATA(self));
```

Comment:
The memory handling in this file ignored `npy_*_cache`

Reply:
Puh, valgrind only helped me to find that a size 1 array was incorrectly added (and which memory address), stepping through with gdb and looking at the cache found this after a long time:

At this point, the size of the array is _already modified_.  That means that `PyArray_NBYTES` should effectively be 0 or so, since the original array was just a dummy defined somewhere else (I did not track it down).  It is fine to add this here, if you calculate the size earlier.  This function is all a bit shady probably, so I am not sure if you should also move the freeing earlier (also NULL'ing directly after the free might be good?  If there is an error, at least we can't run into a double-free).

Reply:
Eh, github showed my comment twice, but it was only there once...

After lots of debugging (first with valgrind, but that only showed me that size 1 (or 0) allocations were added on the cache as larger ones... And annoying, slow and careful stepping through with gdb:

At this point the size of the array is already modified, so freeing (and adding to cache) just doesn't work unless you calculate the size earlier (and maybe move the free to an earlier point). This function is a bit weird, maybe it makes sense to NULL data right away?  Its likely still buggy, but at least we can't run into a double-free then.